storage/kafka/src/main/java/io/apicurio/registry/kafka/KafkaRegistryStorage.java

```diff
@@ -48,102 +63,405 @@
+        }
+
+        Reg.RegistryValue.Builder rvb = getRVBuilder(Reg.ValueType.RULE, actionType, artifactId, -1).setRule(builder);
+        return submit(rvb.build());
+    }
+
+    private <T> CompletableFuture<T> submitSnapshot(long timestamp) {
+        Reg.SnapshotValue.Builder builder = Reg.SnapshotValue.newBuilder().setTimestamp(timestamp);
+        Reg.RegistryValue.Builder rvb = getRVBuilder(Reg.ValueType.SNAPSHOT, Reg.ActionType.CREATE, null, -1).setSnapshot(builder);
+        return submit(rvb.build());
+    }
+
     @SuppressWarnings("unchecked")
-    private <T> CompletableFuture<T> submit(Reg.ActionType actionType, String artifactId, ArtifactType artifactType, long version, String content) {
+    private <T> CompletableFuture<T> submit(Reg.RegistryValue value) {
         UUID reqId = UUID.randomUUID();
-        CompletableFuture<Object> cf = new CompletableFuture<>();
+        CF<Object> cf = new CF<>();
         cfMap.put(reqId, cf);
-        return send(reqId, actionType, artifactId, artifactType, version, content)
+        return send(reqId, value)
             .whenComplete((r, x) -> {
                 if (x != null) cfMap.remove(reqId);
             })
             .thenCompose(r -> (CompletableFuture<T>) cf);
     }
 
-    private CompletableFuture<?> send(UUID reqId, Reg.ActionType actionType, String artifactId, ArtifactType artifactType, long version, String content) {
-        Reg.SchemaValue.Builder builder = Reg.SchemaValue.newBuilder()
-                                                         .setType(actionType)
-                                                         .setArtifactId(artifactId)
-                                                         .setVersion(version);
-        if (artifactType != null) {
-            builder.setArtifactType(artifactType.ordinal());
-        }
-        if (content != null) {
-            builder.setContent(ByteString.copyFrom(content, StandardCharsets.UTF_8));
-        }
-        ProducerRecord<Reg.UUID, Reg.SchemaValue> record = new ProducerRecord<>(
-            KafkaRegistryConfiguration.SCHEMA_TOPIC,
+    private CompletableFuture<?> send(UUID reqId, Reg.RegistryValue value) {
+        ProducerRecord<Reg.UUID, Reg.RegistryValue> record = new ProducerRecord<>(
+            registryTopic,
             ProtoUtil.convert(reqId),
-            builder.build()
+            value
         );
-        return schemaProducer.apply(record);
+        return registryProducer.apply(record);
     }
 
-    public void consumeSchemaValue(ConsumerRecord<Reg.UUID, Reg.SchemaValue> record) {
-        offset = record.offset();
-        Reg.SchemaValue schemaValue = record.value();
-        Reg.ActionType type = schemaValue.getType();
-        CompletableFuture<Object> cf = cfMap.remove(ProtoUtil.convert(record.key()));
+    public void consumeRegistryValue(ConsumerRecord<Reg.UUID, Reg.RegistryValue> record) {
+        this.offset = record.offset();
+
+        boolean isProducer = false;
```

Comment:
Can you explain the mechanics of this? I've got a number of instances of registry and one of them is acting as producer. What happens when the producer crashes (taking its `cfMap` state with it)? Won't that result in the new producer now calling `makeSnapshot(timestamp)` below because its `cfMap` never had the relevent key?

Maybe this is what `ValueType.SNAPSHOT` is for, but I think this is complex enough that it warrants a description of the intended algorithm, because it's not apparent to me as a reader of the code.

Reply:
> I don't think you have to do anything immediately, but one common technique is to fragment the data across multiple Kafka records, and reassemble it in the consumer.

Interesting.
How do you do this?

Reply:
> Won't that result in the new producer now calling makeSnapshot(timestamp) below because its cfMap never had the relevent key?

Just the opposite -- it won't result in a snapshot -- as there would be no matching cf.

My idea was such:
- all nodes check for stale snapshot (with diff delay -- so not all at the same time fire this msg (or highly unlikely))
- the first one to spot this, fires a SNAPSHOT msg
- the producer node makes the snapshot, others just update the lastSnapshotTime

If the producer node crashed in between, there would be no snapshot, yet the other nodes would think there was ...
Hence my asking for the right configuration / ratio between snapshot.period and registry topic retention.
Since it's not a huge problem if we still have good-enough old snapshot.

Or any other idea, how to make this more bullet-proof? :-)

Reply:
How does a producer node get chosen in the first place? And how does a new one get chosen? In particular how long can the delay be between crash of old producer and the availability of a new producer?

Reply:
The first node to "see" that the snapshot is stale issues the SNAPSHOT msg.

> In particular how long can the delay be between crash of old producer and the availability of a new producer?
snapshot.period.minuts

Reply:
I'm not sure I understand yet. 

1. Producer crashes
2. Some other node sees a stale snapshot and issues SNAPSHOT msg
3. But there's no producer to process the SNAPSHOT. Other nodes see SNAPSHOT and update their `lastSnapshotTime`.

The end result is that no snapshot gets taken, AFAICS. Am I missing some detail?

Reply:
The one that issued the SNAPSHOT msg is the producer -- as it has TimedFuture registered in that map.

Reply:
Ah, I think I see now. In essence they're all racing to be the producer of the next snapshot. The first one to see a stale snapshot wins and the race starts again. It's possible that >1 node thinks it won the race, in which case there will be >1 snapshots taken, possibly with different offsets, but that should still be safe (just a waste of persistent storage).

Reply:
Exactly.
Hence that random delay, hoping they won't be concurrent.
But if they are, no big harm done.